### PR TITLE
[@types/bl] update bl types to v5

### PR DESCRIPTION
**Pull request by achingbrain**
The bl module has a file called `BufferList.js` that exports a class `BufferList` that is just the buffer list without node streams, the idea being you can use it in the browser without having to pull in a readable streams implementation, which has become more necessary since webpack 5 stopped bundling shims for node internals.

The main `bl.js` file exports `BufferListStream` that extends both the `BufferList` and the node Duplex stream.

I've attempted to represent the `BufferList.js` class here.

TS doesn't have multiple inheritance so I'm not sure how to represent the `BufferListStream` class without duplicating the methods of `BufferList`.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by achingbrain** [comment]
cc @rvagg maybe these could live the in bl module when the dust settles?

**Comment by rvagg** [comment]
@achingbrain yes, I wasn't even aware this existed until I saw it in one of the recent ipfs PRs! I don't believe anyone tried to get it into `bl` but I'm very interested in solving the double-package pain for something relatively simple.

**Comment by benallfree** [comment]
@rvagg You might remember me from https://github.com/rvagg/bl/pull/99. I also updated these typings a while ago here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51096 as they were pretty out of date. It would be great to get these typings into `bl` just to ensure they stay in sync. If you feel you need any ongoing help with this aspect, I'd be happy to 'own' the TS typings if you don't have anyone else.

@achingbrain Nice work on this PR!

**Comment by benallfree** [comment]
@achingbrain Please see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51096#issuecomment-787427489, I think there may still be a couple methods missing from the `BufferList` definition I previously modified when bringing this up to v4 compat.

**Comment by achingbrain** [comment]
I've added the missing methods and tests for those methods, but https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51096#issuecomment-774550157 refers to the named exports for `BufferListStream` and `BufferList` - I'm not sure how to get TS to express a default export as a class with the extra properties.

I don't think we can just define the classes as static properties of the exported class without TS complaining (despite `esModuleInterop` being turned on):

> This module can only be referenced with ECMAScript imports/exports by turning on the 'esModuleInterop' flag and referencing its default export.ts(2497)

Also attempting to new up instances of the named exports results in:

> This expression is not constructable. Type 'BufferList' has no construct signatures.ts(2351)

Any ideas?  I like the sense of completeness but I'm tempted to not try to solve this problem in this PR, mostly because this is going beyond my ability to contort TS into the shape we need it to be 😉.

**Comment by benallfree** [comment]
@achingbrain Thank you for adding the missing methods!

I sent you a PR https://github.com/achingbrain/DefinitelyTyped/pull/1, see what you think. All the tests seem to pass but yeah I've never had to contend with this either! 👯

File: types/bl/index.d.ts
Comment (benallfree) [reply types/bl/index.d.ts:1]: From the readme: 


> If you intend to continue updating the older version of a library's type declarations, you may create a new subfolder (e.g. /v2/) named for the current (soon to be "old") version, and copy existing files from the current version to it.

Comment (benallfree) [reply types/bl/index.d.ts:1]: @amcasey I'm one of the maintainers of this package; not sure if your suggestion is worthwhile only because @rvagg has expressed interest in moving this into `bl` as a first-class citizen and existing users can pin the @types/bl package version they need in the mean time.

Comment (amcasey) [reply types/bl/index.d.ts:1]: @benallfree Creating an old-version directory is very optional and I was simply asking whether it would be beneficial for this package.  Having said that, I'm not sure I understand how the comment you linked answers the question - first class types would be great but users of v4 would still consume `@types/bl` and might still want bug fixes for those types.  If usage of v4 is limited or if the types are simple enough to be "obviously correct", then it's probably not worthwhile.

Comment (benallfree) [reply types/bl/index.d.ts:1]: @amcasey Sorry yes I meant to say that I see your point and it could be valuable to maintain older versions. There are already known issues/shortcomings with the v4 typings, for example.

I was attempting to balance that against how much longer this package would need to be maintained, but now that I think about it more, I realize that this @types/bl package can really never be deprecated because older versions of bl still need it.

So I agree with you now :) We should folder-ize this package.

**Review by amcasey — CHANGES_REQUESTED** [review]
Sounds like we want to add a v4 folder.

File: types/bl/index.d.ts
Comment (achingbrain) [reply types/bl/index.d.ts:1]: I don't think this is necessary, at least not yet - the bump from v4 to v5 happened because the underlying [buffer](https://www.npmjs.com/package/buffer) browser polyfill jumped from v5 to v6, removing ie11 and Safari < 10 support, not because there was any API breakage.

Would it be better to hold off creating folders until there's a release that breaks the API, give that if we create folders now the contents of v4 and v5 would be identical?

File: types/bl/BufferList.d.ts
Comment (achingbrain) [reply types/bl/BufferList.d.ts:1]: I've removed it

File: types/bl/bl-tests.ts
Comment (achingbrain) [reply types/bl/bl-tests.ts]: No reason, I've reverted it.

File: types/bl/index.d.ts
Comment (benallfree) [reply types/bl/index.d.ts:1]: @achingbrain Hmm I agree with your point too! I'm good either way 👍

**Review by benallfree — CHANGES_REQUESTED** [review]
@achingbrain did you want to incorporate my PR that simplifies the BufferListStream typings? Otherwise I'm good to approve this as is.

**Comment by achingbrain** [comment]
> did you want to incorporate my PR that simplifies the BufferListStream typings?

I'm just trying the changes out on a few projects.  I'd made a similar change initially but it didn't work out, though that may have just been user error..

**Comment by achingbrain** [comment]
@benallfree I've applied your changes, they look good - thanks for the suggestions!

**Review by benallfree — APPROVED** [review]
@achingbrain Thank you for all this awesome work! Looks good for sure now.

File: types/bl/index.d.ts
Comment (amcasey) [reply types/bl/index.d.ts:1]: @achingbrain I'm not sure I understand your suggestion.  I think you're saying that users currently consuming bl@4 should start consuming @types/bl@5 because the APIs are compatible?

**Comment by achingbrain** [comment]
Ready to merge

File: types/bl/index.d.ts
Comment (amcasey) [reply types/bl/index.d.ts:1]: To reiterate though, I'm fine with not creating a v4 directory.

**Comment by amcasey** [comment]
@achingbrain It's fine - it can be added on-demand later.

File: types/bl/index.d.ts
Comment (achingbrain) [reply types/bl/index.d.ts:1]: @amcasey yes, that’s what I’m saying - the APIs are compatible.

**Pull request merged**
